[BUGFIX] Fix handling of failures to fetch the PR list and the PR comments

proski opened the pull request:
```
*  Fix handling of failures to fetch the PR list and the PR comments
   
   Add a new exception StashApiException, throw it in both cases.
   
   Don't trigger a build on a failure to fetch PR comments. That issue has
   been observed creating a large number of builds if the connection between
   Jenkins and Stash is unstable.
```

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:
```diff
@@ -101,9 +102,10 @@ public StashApiClient(
       logger.log(Level.WARNING, "invalid pull request response.", e);
     }
-    return Collections.emptyList();
+    return null;
```
I would rather not use null as special value marker. If we need the calling code to recognize that an error happen why not throw an exception?

proski replied:
Good point. Using custom exceptions would improve the code. I started working on it. It would be a bigger change because it would involve changes to the exception handling and logging.

jakub-bochenski replied:
I think for now we could just throw some generic exceptions, that would still be better than returning null IMO

proski committed:
Fix handling of failures to fetch the PR list and the PR comments

Add a new exception StashApiException, throw it in both cases.

Don't trigger a build on a failure to fetch PR comments. That issue has
been observed creating a large number of builds if the connection between
Jenkins and Stash is unstable.

proski commented:
Rewritten from scratch with a custom exception `StashApiException`.

proski committed:
Add @Nonnull annotations, remove unnecessary null checks

In StashApiClient.java, getPullRequests() and getPullRequestComments()
don't return null anymore, they throw StashApiException on error.
extractComments() is used getPullRequestComments() and doesn't return
null either, so annotate it as well.

In StashRepository.java, remove null checks for the results of the
@Nonnull methods.

proski commented:
Added a separate commit to add Nonnull annotations in the functions that throw instead of returning null. The commit also removes null checks that are not necessary anymore. FindBugs reported one of those checks as a bug.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/StashApiClient.java:
```diff
@@ -562,4 +570,16 @@ public Socket createSocket(Socket socket, String host, int port, boolean autoClo
     }
   }
+
+  public static class StashApiException extends Exception {
```
minor: we could make it into a RuntimeException to avoid having to update all the test signatures

proski replied:
No, I believe a checked exception is more suitable here. Exceptions related to Stash communications should be caught inside the plugin, and not necessarily where they are thrown (so that the plugin can skip building the whole PR if it cannot read just one page of the PR comments). Checked exceptions guarantee that the exception is caught somewhere.

Besides, RuntimeException is for issues that can be prevented programmatically, e.g. for out of bounds access.

The issue is discussed here: https://stackoverflow.com/questions/2190161/

jakub-bochenski replied:
Agree. 

> Exceptions related to Stash communications should be caught inside the plugin, and not necessarily where they are thrown (so that the plugin can skip building the whole PR if it cannot read just one page of the PR comments).

Could you add this to the exceptions javadoc?

proski replied:
I've added a similar text
